File: src/main/java/xyz/nkomarn/composter/network/protocol/packet/serverbound/ServerboundLoginPacket.java
package xyz.nkomarn.composter.network.protocol.packet.serverbound;

import xyz.nkomarn.composter.network.WrappedBuff;
import xyz.nkomarn.composter.network.protocol.handler.PacketHandler;

public class ServerboundLoginPacket implements ServerboundPacket {

    private int protocolVersion;
    private String username;

    public int protocolVersion() {
        return protocolVersion;
    }

    public String username() {
        return username;
    }

    @Override
    public void decode(WrappedBuff buff) {
        protocolVersion = buff.readInt();
        username = buff.readString();
    }

    @Override
    public void handle(PacketHandler handler) {
        handler.handleLogin(this);
    }
}

File: src/main/kotlin/kyta/composter/protocol/packet/login/ServerboundLoginPacket.kt
package kyta.composter.protocol.packet.login

import kyta.composter.protocol.PacketHandler
import kyta.composter.protocol.PacketSerializer
import kyta.composter.protocol.ReadBuffer
import kyta.composter.protocol.ServerboundPacket

data class ServerboundLoginPacket(val protocolVersion: Int, val username: String) : ServerboundPacket {
    override suspend fun handle(handler: PacketHandler) = handler.handleLogin(this)

    companion object : PacketSerializer<ServerboundLoginPacket> {
        override fun deserialize(buffer: ReadBuffer): ServerboundLoginPacket {
            return ServerboundLoginPacket(buffer.readInt(), buffer.readString())
        }
    }
}